ibis/omniscidb/operations.py

```diff
@@ -313,6 +313,90 @@ def extract_field_formatter(translator, expr, sql_attr=sql_attr):
     return extract_field_formatter
 
 
+def _n_weeks_year(year: ir.NumericValue):
+    """
+    Return the number of the total of weeks for the given year.
+
+    Parameters
+    ----------
+    year : ibis.expr.types.NumericValue
+
+    Returns
+    -------
+    Ibis.expr.types.Expr
+
+    See Also
+    --------
+    https://en.wikipedia.org/wiki/ISO_week_date
+    """
+    # n_weeks_year adjustment
+    def _p(year):
+        return (year + (year // 4) - (year // 100) + (year // 400)) % 7
+
+    return 52 + (
+        ibis.case()
+        .when((_p(year) == 4) | (_p(year - 1) == 3), 1)
+        .else_(0)
+        .end()
+    )
+
+
+def _woy_preliminary(d: Union[ir.DateValue, ir.TimestampValue]) -> ir.Expr:
+    """
+    Return a preliminary week of year (WOY).
+
+    Parameters
+    ----------
+    d : ibis.expr.types.DateValue or ibis.expr.types.TimestampValue
+
+    Returns
+    -------
+    ir.Expr
+
+    See Also
+    --------
+    https://en.wikipedia.org/wiki/ISO_week_date
+    """
+    doy = d.day_of_year()
+    dow = d.day_of_week.index() + 1
+
+    result = doy - dow
+    result += 10
+    return (result // 7).cast('int16')
+
+
+def _extract_woy(translator, expr) -> str:
+    def _extract_woy_expr(
+        d: Union[ir.DateValue, ir.TimestampValue]
+    ) -> ir.Expr:
+        """
+        Extract the week of the year expression for a given date/datetime.
+
+        Parameters
+        ----------
+        d : ibis.expr.types.DateValue or ibis.expr.typesTimestampValue]
+
+        Returns
+        -------
+        ir.Expr
+        """
+        w = _woy_preliminary(d)
+        y = d.year()
+
+        one = ibis.literal(1, type='int16')
+
+        return (
+            ibis.case()
+            .when(w < one, _n_weeks_year(y - one))
+            .when(w > _n_weeks_year(y), one)
+            .else_(w)
+            .end()
+        )
+
+    op = expr.op()
+    return translator.translate(_extract_woy_expr(op.args[0]))
```

Comment:
Why do we need a nested function here?

Also, things like `op = expr.op()` seem unnecessary and confusing, if you are going to use the result just once.

Reply:
you're right .. probably I needed that for some reason in the past .. or something like that .. and forgot to refactor that. thanks!